polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -134,13 +133,79 @@ void UniformityAnalysis::visitOperation(
   // A memory side effects free operation that has uniform operands yields
   // uniform result(s).
   if (isMemoryEffectFree(op)) {
-    LLVM_DEBUG(llvm::dbgs() << "Operation is memory effect free\n");
+    LLVM_DEBUG(llvm::dbgs() << "UA: Operation is memory effect free\n");
     return propagateAllIfChanged(results, Uniformity::getUniform());
   }
 
   return analyzeMemoryEffects(op, operands, results);
 }
 
+void UniformityAnalysis::visitNonControlFlowArguments(
+    Operation *op, const RegionSuccessor &successor,
+    ArrayRef<UniformityLattice *> argLattices, unsigned firstIndex) {
+  LLVM_DEBUG(llvm::dbgs() << "UA: Visiting non-ctrl-flow args of operation: "
+                          << *op << "\n");
+
+  // If the lattice on any operand isn't yet initialized, bail out.
+  if (anyOfUniformityIsUninitialized(op->getOperands())) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "UA: Operand(s) uniformity not yet initialized\n");
+    return;
+  }
+
+  // Infer the uniformity of the loop IV by analyzing the loop bounds and step.
+  if (auto loop = dyn_cast<LoopLikeOpInterface>(op)) {
+    std::optional<Value> iv = loop.getSingleInductionVar();
+    if (!iv)
+      return SparseDataFlowAnalysis::visitNonControlFlowArguments(
+          op, successor, argLattices, firstIndex);
+
+    // Collect the bounds and step if they aren't constant.
+    SmallVector<Value> boundsAndStep;
+    Operation *loopOp = loop;
+    if (auto affineLoop = dyn_cast<affine::AffineForOp>(loopOp)) {
+      // Non-constant lower/upper bounds for affine loops use a map so special
+      // case them here. Note that the step is always a constant.
+      ValueRange lbOperands = affineLoop.getLowerBoundOperands();
+      ValueRange ubOperands = affineLoop.getUpperBoundOperands();
+      boundsAndStep.append(lbOperands.begin(), lbOperands.end());
+      boundsAndStep.append(ubOperands.begin(), ubOperands.end());
+    } else {
+      std::optional<OpFoldResult> lb = loop.getSingleLowerBound();
+      std::optional<OpFoldResult> ub = loop.getSingleUpperBound();
+      std::optional<OpFoldResult> st = loop.getSingleStep();
+
```

Comment:
Quite a bit of optionality here -- can you maybe add a comment to explain in which situation the `OpFoldResult` is there, but does not contain a `Value`?

Reply:
Sure. Just answering your question here but I will add a comment to the code as well. 
`LoopLikeOpInterface::getSingleLowerBound` delegates its implementation to operations that implements the `LoopLikeOpInterface`.  Some operations might return a `std::nullopt` and some will return a `OpFoldResult` (like `scf::ForOp`). An `OpFoldResult` is `PointerUnion<Attribute, Value>` so it may be a constant (i.e `Attribute`) or a `Value`. 

In summary `getSingleLowerBound`, `getSingleUpperBound` may:
   - return `nullopt` if the operation has no lower bound to give (for whatever reason), or
   - return a `PointerUnion<Attribute, Value>` which is going to be:
       + a Attribute if, for example, the loop bound is known a compile time, or
        + a Value otherwise.

Reply:
Thanks, got it!